### ISSUE TITLE
Run staticcheck, unused and gosimple at once sharing work

### ISSUE THREAD

**Issue by mvdan** [dominikh/go-tools#39]
Like a gometalinter of sorts, but just for your packages since they could likely share the initial loading work with little effort.

I remember seeing this done (by cockroachdb?) but can't find it now. In any case, it would be useful as a go-gettable main package.

**Comment by mvdan** [comment]
As a side note, I'd like to be able to do this at a higher level so that running a bunch of completely unrelated linters that all use `go/loader` could share/cache much of the initial work. But unless there is some caching done by `go/loader` or some common dependency, that's probably not easy to do.

**Comment by myitcv** [comment]
👍  to the suggestion.

To add something semi-related into the mix. 

We have created a simple tool called `gai` that accepts a list of package specs (i.e. per the `cmd/go` tool). When invoking `gai` you can further specify which of these package specs are "test specs" (explanation follows). Resolving these specs via [`github.com/kisielk/gotool`](https://github.com/kisielk/gotool) gives us two distinct sets (excuse the slight abuse of terminology but just so I can refer to them below):

* Set 1 - a set of packages
* Set 2 - a set of _test_ packages, i.e. the resolved set of test specs

Let us further define:

* Set 3 - the [`TestImports`](https://godoc.org/go/build#Package) of set 2

`gai` then runs `go install` on the union of sets 1, 2 and 3. This step may fail for some subsets of 1, 2 and 3.

`gai` then runs [`gotype`](https://godoc.org/golang.org/x/tools/cmd/gotype) (which uses `go/importer.Default` for speed) as follows:

* `gotype`: those packages from sets 1 and 3 that failed to `go install`
* `gotype -a`: (i.e. include `_test.go` files) all packages from set 2 whose dependencies (including test dependencies) succeeded the `go install` step

This gives us:

* current, i.e. up-to-date, `.a` package files (good for `go vet`, `gocode`, tools that rely on `.a` files)
* extremely fast feedback on compile errors for packages. For those packages that were originally specified as test specs, we also get compile errors from their test files (without needing to run `go test`)

We also know/have:

* which packages (and their tests) out of our original specifications compile
* `*build.Package` values for packages and their transitive dependencies
* the reverse dependencies for all packages

Our idea was to extend `gai` as follows: after outputting to `stderr` the results of the various `gotype` invocations, run a series of vetters/linters on those packages "of interest" that compile (and whose tests compile). Thinking being that:

* we can reuse the `*build.Package` values
* we can parse an `*ast.Package` once per package and re-use that value (and associated `ast.*` values) between vetters/linters
* we can re-use the `*types.Package` that is fed into the `gotype` step (we invoke `gotype` right now but this can trivially be refactored)
* we can also adopt a [`gt`](https://godoc.org/rsc.io/gt) strategy to only re-run vetters/linters if a package (or any of its dependencies) has changed; this will be a big win

Presumably @mvdan we could augment our proposal to also share/cache SSA related values. This is not to suggest that the "main package" you refer to could/should be `gai`; rather just to share the above and get any feedback people may have on our approach. 

Thoughts much appreciated, and either way, look forward to hearing where @mvdan's original suggestion goes.

**Comment by mvdan** [comment]
Thanks for the comment, @myitcv. When I wrote `unless there is some caching done by go/loader or some common dependency`, this is pretty much what I meant. What I understand you would do here is fork and modify linters so that they will conform to an interface of some sort, so you can call them with an `ast.Package` or a `types.Package`.

I personally think that this should be done at another level. If a linter uses `go/loader`, that should be enough for it to be able to reuse type information that has been cached.

In any case, I'd rather not spam this issue which has a much narrower focus. If you'd like to discuss any of this, we should do it elsewhere.

**Comment by myitcv** [comment]
> In any case, I'd rather not spam this issue which has a much narrower focus. If you'd like to discuss any of this, we should do it elsewhere

Good point, apologies for the noise in this issue.

**Comment by mvdan** [comment]
@myitcv if you open up a repo for that tool you were talking about, feel free to open an issue and ping me :)

**Comment by myitcv** [comment]
@mvdan happily - it's on my list of things to do! Would very much appreciate feedback from you, @dominikh and others who work in this space.

**Comment by dominikh** [comment]
@myitcv For now, I'll only provide feedback on the idea of running my tools, and unrelated tools, on the same set of go/loader and go/ssa output: It won't cleanly be possible. One big reason is that we use a custom fork of go/ssa that produces different code than upstream. So already you'd need to generate both forms, and give each tool the correct one. go/loader also has a number of knobs, such as which function bodies to type check, what flags to pass to the source parser, and so on. Different tools may expect different configurations here, so even my tools aside, you may not be able to pass the same program to all tools.

@mvdan cockroachdb has a "meta linter" at https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/metacheck/main.go – it'd be easy for us to provide an official version, the hardest part would be figuring out a suitable name.

**Comment by mvdan** [comment]
> it'd be easy for us to provide an official version

Yes, this is what I meant - thanks for the link.

> the hardest part would be figuring out a suitable name.

Providing an extra binary like `metacheck` could be an option. Perhaps that's a good enough name, as the rest of the tools don't have particularly unique names.

What about structuring it like `go tool`? As in, multiple binaries all shipped in one big binary. Then, you could do stuff like:

```
metacheck all # run all linters
metacheck unused # only unused
metacheck staticcheck # only unused
```

As for the name, I personally think `staticcheck` is not a good name for one of the three tools. `simple` and `unused` are clear, but not `staticcheck`. All of them are static checks. Perhaps the name of the broader tool could be `staticcheck`.

**Comment by dominikh** [comment]
Renaming any of the existing tools, or changing what any of the names stands for, is not an option. Neither is making `metacheck` the only entry point an option. Nor should we have two ways of running single tools, which means no `metacheck staticcheck`.

**Comment by dominikh** [comment]
Closed in aeda00f.

**Comment by mvdan** [comment]
Much appreciated :)
